### PR TITLE
lib/protocol: Understand older/newer Hello messages (fixes #3287)

### LINKED ISSUE #3287: Handle incorrect version hello message / magic
The connection handshake contains

1. A magic number that indicates that this is a Syncthing/BEP connection
2. A "Hello" message that contains some basic metadata about the other side (client name and version, etc).

We should make sure to handle incompatibilities here and display a message that is as friendly and informative as possible.

In particular we should keep track of current and historical magic numbers. This should only change in case the protocol changes in it's entirety, but we know that's going to happen at least once more. Hence we should be able to say not only "incorrect magic" but "The device xyz is using protocol version blah which is not compatible with this release - please upgrade".

If it seems feasible we might also isolate and keep around the code to parse hello messages for old versions. Again this will be less of an issue in the future, but we know one big break is coming. Being able to say with precision that "The device called jb-laptop is running Syncthing v0.13.47 which is too old for this Syncthing v0.14.2" is better than just "some device is too old".

"Planned" as this needs to go into v0.13, and sooner rather than later.

### PR DESCRIPTION
**Pull request by calmh**
### Purpose

This is in preparation for future changes, but also improves the handling when talking to pre-v0.13 clients. It breaks out the Hello message and magic from the rest of the protocol implementation, with the intention that this small part of the protocol will survive future changes.

To enable this, and future testing, the new ExchangeHello function takes an interface that can be implemented by future Hello versions and returns a version indendent result type. It correctly detects pre-v0.13 protocols and returns a "too old" error message which gets logged to the user:

    [I6KAH] 09:21:36 INFO: Failed to exchange Hello messages with [...]:
      the remote device speaks an older version of the protocol (v0.12) not
      compatible with this version

Conversely, something entirely unknown will generate:

    [I6KAH] 09:40:27 INFO: Failed to exchange Hello messages with [...]:
      the remote device speaks an unknown (newer?) version of the protocol

The intention is that in future iterations the Hello exchange will succeed on at least one side and ExchangeHello will return the actual data from the Hello together with ErrTooOld and an even more precise message can be generated.

### Testing

The exchange itself is unit tested. I've tested manually that v0.13 to v0.13 still connect and that the above error is generated when talking to a v0.12 client.

### CODE REVIEW

File: lib/protocol/hello.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+type HelloResult struct {
+	DeviceName    string
+	ClientName    string
+	ClientVersion string
+}
+
+var (
+	// ErrTooOld is returned by ExchangeHello when the other side speaks an
+	// older, incompatible version of the protocol.
+	ErrTooOld = errors.New("the remote device speaks an older version of the protocol (v0.12) not compatible with this version")
```
Comment (AudriusButkevicius) [inline lib/protocol/hello.go]: If we hard code the version, we should call it TooOldV12

File: lib/connections/service.go [hunk 2]
```diff
@@ -153,12 +152,14 @@ next:
 			continue
 		}
 
-		hello, err := exchangeHello(c, s.model.GetHello(remoteID))
+		c.SetDeadline(time.Now().Add(20 * time.Second))
+		hello, err := protocol.ExchangeHello(c, s.model.GetHello(remoteID))
 		if err != nil {
 			l.Infof("Failed to exchange Hello messages with %s (%s): %s", remoteID, c.RemoteAddr(), err)
```
Comment (AudriusButkevicius) [inline lib/connections/service.go]: Probably want a sync.Once warming if TooOld

File: lib/protocol/hello.go (hunk 1 above)
Comment (calmh) [inline lib/protocol/hello.go]: Indeed

File: lib/connections/service.go (hunk 2 above)
Comment (calmh) [inline lib/connections/service.go]: Nah... we print info level message for every other occurrence of something interesting in the connection loop so I think we should do that here too.

Comment (calmh) [inline lib/connections/service.go]: Sorry, I misunderstood. Yes, we do want that.

**Comment by calmh** [comment]
I made it a warning when it's a known version mismatch, but I don't think it's worth rate limiting it. Better to have it front and center and it'll go away once solved.

**Comment by AudriusButkevicius** [comment]
Or just warned map[remoteID]struct{}{}

**Comment by calmh** [comment]
But we want to re-warn now and then or the warning might get eaten by other things like the puller not making progress. Preventing the recurring warning feels like a non-problem to me at the moment?

**Comment by AudriusButkevicius** [comment]
Well with reattempts every minute it might be super annoying. Plus the warnings create popups in the UI, so not sure how you could miss them?

**Comment by calmh** [comment]
We only show/keep the last five warnings IIRC, and the puller can create them on a minutely basis at the least.

**Comment by calmh** [comment]
We could/should rate limit warnings or all log entries based on identical content I guess, but that seems like an overkill solution for a small problem to me at the moment

**Comment by AudriusButkevicius** [comment]
I though puller now shoves stuff in the failed dialog?

**Comment by AudriusButkevicius** [comment]
Nvm it's the final no progress message

**Comment by AudriusButkevicius** [comment]
@st-review merge

**Comment by st-review** [comment]
:ok_hand: Merged as d507126101d019d2511439db0be3b922fd9ff9f3. Thanks, @calmh!

**Comment by calmh** [comment]
No you were actually right, those are just info level. Anyway, lets fix that when someone shows it's an issue :)

**Comment by calmh** [comment]
I've changed my mind, it does get annoying. Will add something reasonable.

**Pull request closed without merging**